File: examples/android/ExampleApp/src/nz/mega/android/bindingsample/BrowserListAdapter.java
/**
 * BrowserListAdapter.java
 * Listview adapter
 *
 * (c) 2013-2014 by Mega Limited, Auckland, New Zealand
 *
 * This file is part of the MEGA SDK - Client Access Engine.
 *
 * Applications using the MEGA API must present a valid application key
 * and comply with the the rules set forth in the Terms of Service.
 *
 * The MEGA SDK is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * @copyright Simplified (2-clause) BSD License.
 *
 * You should have received a copy of the license along with this
 * program.
 */
package nz.mega.android.bindingsample;

import java.text.DecimalFormat;
import java.util.ArrayList;

import nz.mega.sdk.MegaApiAndroid;
import nz.mega.sdk.MegaNode;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

public class BrowserListAdapter extends BaseAdapter {
	
	Context context;
	ArrayList<MegaNode> nodes;
	MegaApiAndroid megaApi;
	
	/* public static view holder class */
	public class ViewHolder {
		RelativeLayout layout;
		ImageView thumbnail;
		TextView fileName;
		TextView fileSize;
	}
		
	BrowserListAdapter (Context context, ArrayList<MegaNode> nodes, MegaApiAndroid megaApi){
		this.context = context;
		this.nodes = nodes;		
		this.megaApi = megaApi;
	}
	
	public void setNodes(ArrayList<MegaNode> nodes){
		this.nodes = nodes;
		notifyDataSetChanged();
	}
	
	@Override
	public int getCount() {
		return nodes.size();
	}

	@Override
	public Object getItem(int position) {
		return nodes.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		
		ViewHolder holder;
		LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		if (convertView == null) {
			convertView = inflater.inflate(R.layout.item_browser_list, parent,false);
			holder = new ViewHolder();
			
			holder.layout = (RelativeLayout) convertView.findViewById(R.id.item_layout);
			holder.thumbnail = (ImageView) convertView.findViewById(R.id.thumbnail);
			holder.fileName = (TextView) convertView.findViewById(R.id.filename);
			holder.fileSize = (TextView) convertView.findViewById(R.id.filesize);
			
			convertView.setTag(holder);
		}
		else{
			holder = (ViewHolder) convertView.getTag();
		}
		
		MegaNode node = (MegaNode) getItem(position);
		holder.fileName.setText(node.getName());
		
		if (node.isFile()){
			holder.fileSize.setText(getSizeString(node.getSize()));
			holder.thumbnail.setImageResource(R.drawable.mime_file);
		}
		else{
			holder.thumbnail.setImageResource(R.drawable.mime_folder);
			holder.fileSize.setText(getInfoFolder(node));
		}
		
		return convertView;
	}

	private String getSizeString(long size){
		String sizeString = "";
		DecimalFormat decf = new DecimalFormat("###.##");

		float KB = 1024;
		float MB = KB * 1024;
		float GB = MB * 1024;
		float TB = GB * 1024;
		
		if (size < KB){
			sizeString = size + " B";
		}
		else if (size < MB){
			sizeString = decf.format(size/KB) + " KB";
		}
		else if (size < GB){
			sizeString = decf.format(size/MB) + " MB";
		}
		else if (size < TB){
			sizeString = decf.format(size/GB) + " GB";
		}
		else{
			sizeString = decf.format(size/TB) + " TB";
		}
		
		return sizeString;
	}
	
	private String getInfoFolder(MegaNode n) {
		ArrayList<MegaNode> nL = megaApi.getChildren(n);

		int numFolders = 0;
		int numFiles = 0;

		for (int i = 0; i < nL.size(); i++) {
			MegaNode c = nL.get(i);
			if (c.isFolder()) {
				numFolders++;
			} else {
				numFiles++;
			}
		}

		String info = "";
		if (numFolders > 0) {
			info = numFolders
					+ " "
					+ context.getResources().getQuantityString(
							R.plurals.general_num_folders, numFolders);
			if (numFiles > 0) {
				info = info
						+ ", "
						+ numFiles
						+ " "
						+ context.getResources().getQuantityString(
								R.plurals.general_num_files, numFiles);
			}
		} else {
			info = numFiles
					+ " "
					+ context.getResources().getQuantityString(
							R.plurals.general_num_files, numFiles);
		}

		return info;
	}
}

File: examples/android/ExampleApp/app/src/main/java/nz/mega/android/bindingsample/BrowserListAdapter.kt
/**
 * BrowserListAdapter.kt
 * Listview adapter
 *
 * (c) 2013-2014 by Mega Limited, Auckland, New Zealand
 *
 * This file is part of the MEGA SDK - Client Access Engine.
 *
 * Applications using the MEGA API must present a valid application key
 * and comply with the the rules set forth in the Terms of Service.
 *
 * The MEGA SDK is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * @copyright Simplified (2-clause) BSD License.
 *
 * You should have received a copy of the license along with this
 * program.
 */
package nz.mega.android.bindingsample

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import nz.mega.sdk.MegaApiAndroid
import nz.mega.sdk.MegaNode
import java.text.DecimalFormat
import java.util.ArrayList

class BrowserListAdapter(
    private val context: Context,
    private var nodes: ArrayList<MegaNode>,
    private val megaApi: MegaApiAndroid
) : BaseAdapter() {

    /* public static view holder class */
    class ViewHolder {
        lateinit var layout: RelativeLayout
        lateinit var thumbnail: ImageView
        lateinit var fileName: TextView
        lateinit var fileSize: TextView
    }

    fun setNodes(nodes: ArrayList<MegaNode>) {
        this.nodes = nodes
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return nodes.size
    }

    override fun getItem(position: Int): Any {
        return nodes[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = convertView ?: run {
            val newView = inflater.inflate(R.layout.item_browser_list, parent, false)
            val newHolder = ViewHolder()
            newHolder.layout = newView.findViewById(R.id.item_layout)
            newHolder.thumbnail = newView.findViewById(R.id.thumbnail)
            newHolder.fileName = newView.findViewById(R.id.filename)
            newHolder.fileSize = newView.findViewById(R.id.filesize)
            newView.tag = newHolder
            newView
        }

        holder = view.tag as ViewHolder

        val node = getItem(position) as MegaNode
        holder.fileName.text = node.name

        if (node.isFile) {
            holder.fileSize.text = getSizeString(node.size)
            holder.thumbnail.setImageResource(R.drawable.mime_file)
        } else {
            holder.thumbnail.setImageResource(R.drawable.mime_folder)
            holder.fileSize.text = getInfoFolder(node)
        }

        return view
    }

    private fun getSizeString(size: Long): String {
        val decf = DecimalFormat("###.##")

        val KB = 1024f
        val MB = KB * 1024
        val GB = MB * 1024
        val TB = GB * 1024

        return when {
            size < KB -> "$size B"
            size < MB -> "${decf.format(size / KB)} KB"
            size < GB -> "${decf.format(size / MB)} MB"
            size < TB -> "${decf.format(size / GB)} GB"
            else -> "${decf.format(size / TB)} TB"
        }
    }

    private fun getInfoFolder(n: MegaNode): String {
        val nL = megaApi.getChildren(n)

        var numFolders = 0
        var numFiles = 0

        for (c in nL) {
            if (c.isFolder) {
                numFolders++
            } else {
                numFiles++
            }
        }

        val info = when {
            numFolders > 0 -> {
                val folderInfo = "$numFolders ${context.resources.getQuantityString(R.plurals.general_num_folders, numFolders)}"
                if (numFiles > 0) {
                    "$folderInfo, $numFiles ${context.resources.getQuantityString(R.plurals.general_num_files, numFiles)}"
                } else {
                    folderInfo
                }
            }
            else -> "$numFiles ${context.resources.getQuantityString(R.plurals.general_num_files, numFiles)}"
        }

        return info
    }
}

